src/twisted/test/test_tcp.py

```diff
@@ -1427,11 +1433,15 @@ def testWriter(self):
         reactor.connectTCP("127.0.0.1", n, wrappedClientF)
 
         d = defer.gatherResults([wrappedF.deferred, wrappedClientF.deferred])
+
         def check(ignored):
             self.assertTrue(f.done, "writer didn't finish, it probably died")
-            self.assertTrue(clientF.len == self.datalen,
+            self.assertTrue(clientF.len >= self.datalen,
                             "client didn't receive all the data it expected "
                             "(%d != %d)" % (clientF.len, self.datalen))
+            self.assertTrue(clientF.len <= self.datalen,
```

Comment:
What is the point of this assertion?
Why the previous one was not good? :)

Reply:
Current IOCP reactor shows rather interesting behaviour. ~10% of runs are ok, ~80% will loose data, and the remaining ~10% return duplicate data. Of course a simple `==` assertion would do, but the error message should not state data is missing.

Reply:
maybe the assertion message can be changed from `didn't received all the data` to `didn't receive the expected data` ... and replace with the dedicated self.assertEqual 

but this is minor comment ...

what I would like to see, is this tests executed without the patch applied and see that it constantly fails
and then when applying the patch we have a solid green test :)

Reply:
Making test fail with probability 1 on the unpatched IOCP reactor will be hard  IMHO. The issue is with the interaction of twisted with the Windows network stack. You had to mock that to get deterministic behaviour. You could run the test multiple times which would obviously reduce chances of false passes.

That said, [this](https://github.com/asodeur/twisted/commit/ca7e96360a1de70cc364c6eb5b0c0b21d75ca4df) is the commit that changed the test. Everything else (including the IOCP reactor) is still unchanged. You might give that a spin to get an idea of the frequency of false passes ...

Reply:
A mock that emulates the behavior of the IOCP APIs under their various failure modes *would* be a welcome addition, especially to help other maintainers who may want to clean things up but don't know as much about IOCP.  That said, it sounds like that would be a good followup.